src/computability/language.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fox Thomson.
+-/
+import data.finset.basic
+
+/-!
+# Languages
+This file contains the definition and operations on formal languages over an alphabet. Note strings
+are implemented as lists over the alphabet.
+-/
+
+universe u
+
+variables {α : Type u} [dec : decidable_eq α]
+
+/-- A language is a set of strings over an alphabet. -/
+@[derive [has_union, has_emptyc, has_mem (list α), has_singleton (list α)]]
+def language (α) := set (list α)
+
+namespace language
+
+instance : inhabited (language α) := ⟨∅⟩
+
+instance : has_zero (language α) := ⟨∅⟩
+instance : has_one (language α) := ⟨{[]}⟩
+
+instance : has_add (language α) := ⟨set.union⟩
+instance : has_mul (language α) := ⟨λ l m, (l.prod m).image (λ p, p.1 ++ p.2)⟩
+
+@[simp] lemma zero_def : (0 : language α) = ∅ := rfl
+@[simp] lemma one_def : (1 : language α) = {[]} := rfl
+
+@[simp] lemma add_def (l m : language α) : l + m = l ∪ m := rfl
+@[simp] lemma mul_def (l m : language α) : l * m = (l.prod m).image (λ p, p.1 ++ p.2) := rfl
+
+/-- The star of a language `L` is the set of all strings which can be written by concatenating
+  strings from `L`. -/
+def star (l : language α) : language α :=
+{ x | ∃ S : list (list α), x = S.join ∧ ∀ y ∈ S, ¬(list.empty y) ∧ y ∈ l}
```

Comment:
```suggestion
{ x | ∃ S : list (list α), x = S.join ∧ ∀ y ∈ S, y ∈ l}
```
You should use `y ≠ []` in preference to `¬(list.empty y)`, but in this case I think you don't need the conjunct at all in the definition. This version of the definition can be a lemma.

Reply:
Out of curiosity, why is `y ≠ []` preferred over `¬(list.empty y)`?

Reply:
the more "mathematical" expressions have more lemmas defined for them. TBH functions like `list.empty` look like a haskellism to me; we use decidable propositions most of the time over boolean predicates, which gives us the advantage of being able to use logical manipulation tactics like `cases` and `intro` instead of inductive proofs all the time. For `list.empty` the difference isn't too big, but for example if you want to prove `list.empty l -> P l` you can use `rintro rfl` if it is written `l = []` but you need `cases l; rintro rfl` for `list.empty`.

Reply:
Thanks, that's helpful!